File: android/gemcore/src/main/kotlin/com/gemwallet/android/domains/perpetual/PerpetualConfig.kt
package com.gemwallet.android.domains.perpetual

import com.gemwallet.android.ext.toAssetId
import com.wallet.core.primitives.AssetId
import uniffi.gemstone.Config
import java.math.BigInteger

object PerpetualConfig {
    private val config get() = Config().getPerpetualConfig()

    val defaultLeverage: Int get() = config.defaultLeverage.toInt()

    val depositAddress: String get() = config.depositAddress

    val depositAssetId: AssetId get() = requireNotNull(config.depositAssetId.toAssetId()) {
        "Invalid perpetual deposit asset id: ${config.depositAssetId}"
    }

    val minDeposit: BigInteger get() = config.minDeposit.toLong().toBigInteger()

    val minWithdraw: BigInteger get() = config.minWithdraw.toLong().toBigInteger()

    val pricesUpdateIntervalSeconds: Int get() = config.pricesUpdateIntervalSeconds.toInt()

    val leverageOptions: List<Int> get() = config.leverageOptions.toUnsignedInts()

    val takeProfitOptions: List<Int> get() = config.takeProfitPercentOptions.toUnsignedInts()

    val stopLossOptions: List<Int> get() = config.stopLossPercentOptions.toUnsignedInts()

    val defaultTakeProfit: Int get() = config.defaultTakeProfitPercent.toInt()

    val defaultStopLoss: Int get() = config.defaultStopLossPercent.toInt()

    fun autocloseSuggestions(leverage: Int): List<Int> =
        Config().getAutocloseSuggestions(leverage.toUByte()).toUnsignedInts()

    fun selectLeverage(desired: Int, from: List<Int>): Int =
        Config().selectLeverage(
            desired.toUByte(),
            from.map { it.toByte() }.toByteArray(),
        ).toInt()
}

private fun ByteArray.toUnsignedInts(): List<Int> = map { it.toUByte().toInt() }

File: ios/Packages/GemstonePrimitives/Sources/PerpetualConfig.swift
// Copyright (c). Gem Wallet. All rights reserved.

import BigInt
import Foundation
import class Gemstone.Config
import Primitives

public struct PerpetualConfig {
    private init() {}

    public static var defaultLeverage: UInt8 {
        Config.shared.getPerpetualConfig().defaultLeverage
    }

    public static var depositAddress: String {
        Config.shared.getPerpetualConfig().depositAddress
    }

    public static var depositAssetId: String {
        Config.shared.getPerpetualConfig().depositAssetId
    }

    public static var depositAsset: Asset {
        guard let assetId = try? AssetId(id: depositAssetId) else {
            preconditionFailure("Invalid perpetual deposit asset id: \(depositAssetId)")
        }
        let usdc = Asset.hypercoreUSDC()
        return Asset(id: assetId, name: usdc.name, symbol: usdc.symbol, decimals: usdc.decimals, type: .token)
    }

    public static var minDeposit: BigInt {
        BigInt(Config.shared.getPerpetualConfig().minDeposit)
    }

    public static var minWithdraw: BigInt {
        BigInt(Config.shared.getPerpetualConfig().minWithdraw)
    }

    public static var pricesUpdateIntervalSeconds: TimeInterval {
        TimeInterval(Config.shared.getPerpetualConfig().pricesUpdateIntervalSeconds)
    }

    public static var leverageOptions: [UInt8] {
        Array(Config.shared.getPerpetualConfig().leverageOptions)
    }

    public static var takeProfitOptions: [UInt8] {
        Array(Config.shared.getPerpetualConfig().takeProfitPercentOptions)
    }

    public static var stopLossOptions: [UInt8] {
        Array(Config.shared.getPerpetualConfig().stopLossPercentOptions)
    }

    public static func selectLeverage(desired: UInt8, options: [UInt8]) -> UInt8 {
        Config.shared.selectLeverage(desired: desired, options: Data(options))
    }

    public static func autocloseSuggestions(leverage: UInt8) -> [UInt8] {
        Array(Config.shared.getAutocloseSuggestions(leverage: leverage))
    }
}
